src/shell/shell_cmds.cpp

```diff
@@ -1390,15 +1468,26 @@ void DOS_Shell::CMD_DATE(char * args) {
 	}
 	// check if a date was passed in command line
 	uint32_t newday, newmonth, newyear;
-	if (sscanf(args, "%u-%u-%u", &newmonth, &newday, &newyear) == 3) {
-		reg_cx = static_cast<uint16_t>(newyear);
-		reg_dh = static_cast<uint8_t>(newmonth);
-		reg_dl = static_cast<uint8_t>(newday);
-
-		reg_ah = 0x2b; // set system date
-		CALLBACK_RunRealInt(0x21);
-		if (reg_al == 0xff) {
+	int n;
+	if (date_format == 1)
+		n = sscanf(args, "%u%c%u%c%u", &newday, &c1, &newmonth, &c2, &newyear);
+	else if (date_format == 2)
+		n = sscanf(args, "%u%c%u%c%u", &newyear, &c1, &newmonth, &c2, &newday);
+	else
+		n = sscanf(args, "%u%c%u%c%u", &newmonth, &c1, &newday, &c2, &newyear);
+	if (n == 5 && c1 == date_separator && c2 == date_separator) {
+		if (!is_date_valid(newyear, newmonth, newday))
 			WriteOut(MSG_Get("SHELL_CMD_DATE_ERROR"));
+		else {
+			reg_cx = static_cast<uint16_t>(newyear);
+			reg_dh = static_cast<uint8_t>(newmonth);
+			reg_dl = static_cast<uint8_t>(newday);
```

Comment:
these are 32-bit values coming from scanf, but I'm not seeing bounds checking to really limit the values to these small type-sizes (so we could get wrapping). 
Can we feed these values through your previous date validation function? 
(if not, a slightly worse option is to use `check_cast` instead of `static_cast`, which does bounds-checking.

Reply:
The is_date_valid function does validate values right before this. I assume it has pretty good checks for the years/months/days in this case. This code is similar to the existing code in the CMD_TIME function.

Reply:
roger that - thanks for clarifying @Wengier !
(resolving)